### ISSUE TITLE
Restricted master is feasible but CPLEX did not return a dual solution

### ISSUE THREAD

**Issue by guilhemdupuis** [atoptima/Coluna.jl#315]
Using CPLEX on the generalized assignment problem demo provided on GitHub, I get an error stating that:
> Solver returned that the LP restricted master is feasible but did not return a dual solution

Here is the log from Coluna:
``` 
<it=844> <et=384.37> <mst= 0.02> <sp= 0.63> <cols= 0> <DB= 1929.6667> <mlp= 1929.6667> <PB=Inf>
[ LogLevel(1): Column Generation Algorithm has converged.
[ Info: Recording reformulation state.
[ Info: Recording master info.
[ Info: Recording sp 4 info.
[ Info: Recording sp 2 info.
[ Info: Recording sp 3 info.
[ Info: Recording sp 5 info.
[ Info: Recording sp 6 info.
[ Info: Updating tree.
Child nodes generated :
***************************************************************************************
**** BaB tree node N° 2, parent N° 1, depth 1, 2 open nodes
**** Local DB = 1929.6667, global bounds : [ 1929.6667 , 1960.0000 ], time = 393.31 sec.
**** Branching constraint: x[1,6]>=1.0
***************************************************************************************
[ Info: Preparing reformulation according to node record
[ Info: Preparing reformulation master
[ Info: Resetting sp 4 state.
[ Info: Resetting sp 2 state.
[ Info: Resetting sp 3 state.
[ Info: Resetting sp 5 state.
[ Info: Resetting sp 6 state.
┌ Error: Solver returned that the LP restricted master is feasible but did not return a dual solution. Please open an issue (https://github.com/atoptima/Coluna.jl/issues).
└ @ Coluna.Algorithm ~/.julia/packages/Coluna/OqFqT/src/Algorithm/colgen.jl:392
<it=  1> <et=395.77> <mst= 0.34> <sp= 0.30> <cols= 0> <DB= 1929.8571> <mlp= 1929.8571> <PB=1960.0000>
[ LogLevel(1): Column Generation Algorithm has converged.
[ Info: Recording reformulation state.
[ Info: Recording master info.
[ Info: Recording sp 4 info.
[ Info: Recording sp 2 info.
[ Info: Recording sp 3 info.
[ Info: Recording sp 5 info.
[ Info: Recording sp 6 info.
[ Info: Updating tree.
```
In addition, we wonder why is the dual solution needed for the restricted master.

**Sources:**
Model and dataset can be found at
https://covid19.framadrop.org/r/6vFxjdD226#Hqayehtwz0hbT0vTiCDmkh9vyLyWC6f1ouBURYgRJBM=
(with GLPK replaced by CPLEX in ColunaDemos.jl)

**Environment:**
- Julia => 1.3.1
- "Coluna" => v"0.3.0"
- "GLPK" => v"0.12.1"
- "BlockDecomposition" => v"1.2.0"
- "MathOptInterface" => v"0.9.13"
- "JuMP" => v"0.21.2"
-  "CPLEX"              => v"0.6.3"
- OS: MacOs

Best regards,

Guilhem

**Comment by guimarqu** [comment]
Same as https://github.com/atoptima/Coluna.jl/issues/302#issuecomment-601720643

You need the dual of the LP restricted master to compute the reduced cost of subproblem variables. (reduced costs = c - yA  where c is the original costs, y the dual solution, A the coeff matrix of the master)

**Comment by guilhemdupuis** [comment]
Thank you for the answer.

This error occurs only when the restricted master IP heuristic is activated. I suppose reduced costs are deduced during column generation otherwise, when there is no restricted master heuristic.

Then, what is the difference between the resolution of the LP restricted master and the last iteration of the column generation ?

**Comment by guimarqu** [comment]
Ok maybe one variable stays integer in the master (related to #314 ?) after the call to the IP restricted master heuristic when the formulation is relaxed. It thus can explain why there is no dual solution.

The LP restricted master is solved at each iteration of column generation. We compute reduced cost of subproblem variables using the dual solution to the LP restricted master.

**Comment by guimarqu** [comment]
I can't reproduce the bug, what is the commit you're working on ?

**Comment by guilhemdupuis** [comment]
I am working with Coluna's master branch, added with`pkg> add Coluna#master`. I can't find the commit sha on my computer but I suppose it is the last one, ie `f0c0853c65fd890cfb8add4791ee66cc7555d8ff`. Anyway, the error also occur when I add Coluna with `pkg> add Coluna#f0c0853c65fd890cfb8add4791ee66cc7555d8ff`.

Thanks for the explanation about the LP restricted master. Because of the position of the error I thought it was caused by the IP restricted master heuristic but I did not understand first that it occured during column generation.
